File: web/src/api/entities.ts
import { apiFetch } from './client'

export interface Entity {
  id: string
  name: string
  base_currency: string
  wallet_count: number
  created_at: string
  updated_at: string
}

export interface EntityList {
  entities: Entity[]
  total: number
}

export interface EntityCreateRequest {
  name: string
  base_currency?: string
}

export interface EntityUpdateRequest {
  name?: string
  base_currency?: string
}

export async function listEntities(): Promise<EntityList> {
  return apiFetch<EntityList>('/entities')
}

export async function createEntity(data: EntityCreateRequest): Promise<Entity> {
  return apiFetch<Entity>('/entities', {
    method: 'POST',
    body: JSON.stringify(data),
  })
}

export async function getEntity(id: string): Promise<Entity> {
  return apiFetch<Entity>(`/entities/${id}`)
}

export async function updateEntity(id: string, data: EntityUpdateRequest): Promise<Entity> {
  return apiFetch<Entity>(`/entities/${id}`, {
    method: 'PATCH',
    body: JSON.stringify(data),
  })
}

export async function deleteEntity(id: string): Promise<void> {
  return apiFetch<void>(`/entities/${id}`, { method: 'DELETE' })
}

File: src/cryptotax/api/schemas/entities.py
import uuid
from datetime import datetime

from pydantic import BaseModel


class EntityCreateRequest(BaseModel):
    name: str
    base_currency: str = "VND"


class EntityUpdateRequest(BaseModel):
    name: str | None = None
    base_currency: str | None = None


class EntityResponse(BaseModel):
    id: uuid.UUID
    name: str
    base_currency: str
    wallet_count: int = 0
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class EntityListResponse(BaseModel):
    entities: list[EntityResponse]
    total: int
